File: src/chainsaddiction/utils/tests/test-utils.c
#include <stdio.h>
#include "test-utils.h"

bool test__local_decoding (void)
{
#ifdef datapath
#undef datapath
#endif
#define datapath "../../../../tests/data/earthquakes/3s/"
    enum {
        m_states = 3,
        n_obs    = 107
    };

    bool     err  = true;
    size_t  *dec  = NULL;
    DataSet *lcsp = NULL;
    DataSet *xpct = NULL;

    lcsp = DataSet_NewFromFile (datapath "lcsp.txt");
    xpct = DataSet_NewFromFile (datapath "local-decoding.txt");
    dec = VA_SIZE_EMPTY (n_obs);
    if (lcsp == NULL || xpct == NULL || dec == NULL)
    {
        DataSet_Delete (lcsp);
        DataSet_Delete (xpct);
        FREE (dec);
        return UT_FAILURE;
    }
    local_decoding (n_obs, m_states, lcsp->data, dec);

    for (size_t i = 0; i < n_obs; i++)
    {
        err = ((size_t)xpct->data[i] != dec[i]) ? true : false;
        if (err) break;
    }

    DataSet_Delete (lcsp);
    DataSet_Delete (xpct);
    free (dec);
    return err;
}


bool test__global_decoding (void)
{
#ifdef datapath
#undef datapath
#endif
#define datapath "../../../../tests/data/earthquakes/3s/"

    enum {
        m_states = 3,
        n_obs    = 107
    };
    bool    err     = false;
    size_t  *dec    = NULL;
    DataSet *lcsp   = NULL;
    DataSet *lgamma = NULL;
    DataSet *ldelta = NULL;
    DataSet *xpct   = NULL;

    lcsp   = DataSet_NewFromFile (datapath "lcsp.txt");
    lgamma = DataSet_NewFromFile (datapath "lgamma.txt");
    ldelta = DataSet_NewFromFile (datapath "ldelta.txt");
    xpct   = DataSet_NewFromFile (datapath "global-decoding.txt");
    dec    = VA_SIZE_EMPTY (n_obs);
    if (lcsp == NULL || lgamma == NULL || ldelta == NULL || 
        xpct == NULL || dec == NULL)
    {
        DataSet_Delete (lcsp);
        DataSet_Delete (lgamma);
        DataSet_Delete (ldelta);
        DataSet_Delete (xpct);
        free (dec);
        return UT_FAILURE;
    }    

    global_decoding (n_obs, m_states, lgamma->data, ldelta->data,
                     lcsp->data, dec);

    for (size_t i = 0; i < n_obs; i++)
    {
        err = ((size_t)xpct->data[i] != dec[i]) ? true : false;
        if (err) break;
    }

    DataSet_Delete (lcsp);
    DataSet_Delete (lgamma);
    DataSet_Delete (ldelta);
    DataSet_Delete (xpct);
    free (dec);
    return err;
}

File: tests/test-utils.py
from pathlib import Path
import sys
import unittest

from utils import load_data
import numpy as np
from chainsaddiction import utils


class TestLocalDecoding(unittest.TestCase):
    def setUp(self):
        src = 'data/earthquakes/3s'
        self.lcxpt = load_data(src, 'lcsp.txt').reshape(-1, 3)
        self.res = load_data(src, 'local-decoding.txt', 'uint64')

    def test_local_decoding(self):
        dec = utils.local_decoding(self.lcxpt)
        self.assertTrue(np.array_equal(dec, self.res))


class TestGloblaDecoding(unittest.TestCase):
    def setUp(self):
        src = 'data/earthquakes/3s'
        self.lcxpt = load_data(src, 'lcsp.txt').reshape(-1, 3)
        self.lgamma = load_data(src, 'lgamma.txt').reshape(3, 3)
        self.ldelta = load_data(src, 'ldelta.txt')
        self.res = load_data(src, 'global-decoding.txt', 'uint64')

    def test_global_decoding(self):
        dec = utils.global_decoding(self.lgamma, self.ldelta, self.lcxpt)
        self.assertTrue(np.array_equal(dec, self.res))


if __name__ == '__main__':
    unittest.main()
